source/slang/slang-ir-spirv-legalize.cpp

```diff
@@ -1167,6 +1167,125 @@ struct SPIRVLegalizationContext : public SourceEmitterBase
         addUsersToWorkList(newStore);
     }
 
+    void processNonUniformResourceIndex(IRInst* nonUniformResourceIndexInst)
+    {
+        // implement the translation to spirv by walking up the use-def chain
+        // from nonUniformResource inst of an index to an array of buffer or
+        // texture def all the way to the leaf operations. To be precise:
+        // - go through GEP and see if it calls an intrinsic function,
+        //   then decorate the address itself (GetElementPtr)
+        // - go through GEP to identify the pointer access and the Loads that it
+        //   accesses (GetElementPtr -> Load), then decorate the load instruction.
+        // - go through IntCasts to deal with u32 -> i32 / vice-versa (IntCast)
+        List<IRInst*> resWorkList;
+
+        // Handle cases when `nonUniformResourceIndexInst` inst is wrapped around
+        // an index in a nested fashion, i.e. nonUniform(nonUniform((index)) by
+        // only adding the inner-most inst in the worklist, and work our way out.
+        auto insti = nonUniformResourceIndexInst;
+        while (insti->getOp() == kIROp_NonUniformResourceIndex)
+        {
+            if (resWorkList.getCount() != 0)
+                resWorkList.removeLast();
+            resWorkList.add(insti);
+            insti = insti->getOperand(0);
+        }
```

Comment:
Would this handle a cases like this?
```
nonUniform(nonUniform(index)+1)
```
I am not sure if it is a valid expression or not.

Reply:
seems legal to me, but it is not currently handled by the change.
this will look something like a 
```
GEP(Buffer, NonUniform(Add(NonUniform(Swizzle(Load, Int)), Int)) )
```

In the current state of this patch, the inner most nonUniform is not handled. @csyonghe do we need to address this?